### PR TITLE
#23429 wrap url rewrite regeneration into transaction to keep existing urls form deletion

### PR DESCRIPTION
**Pull request by ilnytskyi**
If an error occurred, deadlock, exception during URL Rewrite regeneration then old urls are lost and new ones not generated that leads to 404 error page on affected products. 
### Description (*)

Added transaction to rollback unsuccessful regeneration. 

### Fixed Issues (if relevant)

1. magento/magento2#23429: Url Rewrites are lost if an exception thrown or deadlock during regeneration

### Manual testing scenarios (*)

1. Emulate deadlock on `url_rewrite` table 
2. Regenerate URL rewrites 

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/UrlRewrite/Model/Storage/DbStorage.php [hunk 1]
```diff
@@ -258,14 +260,17 @@ private function prepareUniqueEntities(array $urls): array
...
 
     /**
      * @inheritDoc
      */
-    protected function doReplace(array $urls)
+    protected function doReplace(array $urls): array
     {
+        $this->connection->beginTransaction();
+
         $this->deleteOldUrls($urls);
```
Comment (kandy) [inline app/code/Magento/UrlRewrite/Model/Storage/DbStorage.php]: if the exception will throw on this level, the transaction will not be rollback properly
please move "try" statement  directly after the transaction begin statement

**Review by kandy — APPROVED** [review]

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-5446](https://jira.corp.magento.com/browse/ENGCOM-5446) has been created to process this Pull Request
:eight_spoked_asterisk: @sidolov, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
Hi @ilnytskyi.  I have difficulty with reproducing this issue on 2.3-develop.
**Manual testing scenario:**
- Add exception before [function](https://github.com/magento/magento2/blob/7ffabd074c457edb7a3f2d8eacbc21a1dd188692/app/code/Magento/UrlRewrite/Model/Storage/DbStorage.php#L276) `$this->insertMultiple($data)` in `app/code/Magento/UrlRewrite/Model/Storage/DbStorage.php`.
- Edit product URL Key and save it

**Result:**
:heavy_check_mark: Exception is caught by [EntitiyManager](https://github.com/magento/magento2/blob/7ffabd074c457edb7a3f2d8eacbc21a1dd188692/lib/internal/Magento/Framework/EntityManager/Operation/Update.php#L97) that does transaction [rollback](https://github.com/magento/magento2/blob/7ffabd074c457edb7a3f2d8eacbc21a1dd188692/lib/internal/Magento/Framework/EntityManager/Operation/Update.php#L123) 
:heavy_check_mark: URL leads to correct product page

**Comment by ilnytskyi** [comment]
@engcom-Delta 
My problem was that `$this->deleteOldUrls($urls);` removed old previous urls, and 
`$this->insertMultiple($data);` did not not generate new if an exception thrown so I decided to wrap both into transaction. In case of anything, it would revert old urls.

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-5446](https://jira.corp.magento.com/browse/ENGCOM-5446) has been created to process this Pull Request
:eight_spoked_asterisk: @VladimirZaets, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Pull request merged**